File: app/src/main/java/com/metahumanz/pacilread/reader/modern/ui/ReaderStyleController.java
package com.metahumanz.pacilread.reader.modern.ui;

import android.content.Intent;
import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.graphics.Color;
import android.graphics.Typeface;
import android.net.Uri;
import android.os.Build;
import android.view.Gravity;
import android.view.View;
import android.view.ViewGroup;
import android.view.WindowManager;
import android.widget.TextView;

import com.metahumanz.pacilread.reader.modern.ModernReaderActivity;
import com.metahumanz.pacilread.reader.modern.ReaderRuntime;
import com.metahumanz.pacilread.reader.modern.ReaderSessionState;
import com.metahumanz.pacilread.reader.modern.ReaderUiUtils;
import com.metahumanz.pacilread.reader.modern.ReaderViewRefs;
import com.metahumanz.pacilread.reader.modern.config.ReaderOptionCatalog;
import com.metahumanz.pacilread.reader.modern.content.ReaderContentController;
import com.metahumanz.pacilread.reader.modern.paging.ReaderPagingAnimator;
import com.metahumanz.pacilread.reader.modern.theme.ReaderDisplayModeHelper;
import com.metahumanz.pacilread.reader.modern.theme.ReaderThemePalette;
import com.metahumanz.pacilread.tts.MimoTtsClient;
import com.metahumanz.pacilread.util.FileAssetHelper;

import java.io.File;
import java.util.Locale;

public final class ReaderStyleController {
    private static final float CHAPTER_TITLE_SCALE = 1.4f;
    private static final int HUD_BAR_HEIGHT_DP = 22;
    private static final float HUD_TEXT_SIZE_SP = 12f;
    private static final float HUD_TEXT_ALPHA = 0.78f;

    private final ModernReaderActivity activity;
    private final ReaderRuntime runtime;
    private final ReaderViewRefs views;
    private final ReaderSessionState state;
    private final ReaderUiUtils ui;

    private ReaderChromeController chrome;
    private ReaderPagingAnimator paging;
    private ReaderContentController content;
    private com.metahumanz.pacilread.reader.modern.tts.ReaderTtsController tts;

    public ReaderStyleController(
            ModernReaderActivity activity,
            ReaderRuntime runtime,
            ReaderViewRefs views,
            ReaderSessionState state,
            ReaderUiUtils ui
    ) {
        this.activity = activity;
        this.runtime = runtime;
        this.views = views;
        this.state = state;
        this.ui = ui;
    }

    public void attachControllers(
            ReaderChromeController chrome,
            ReaderPagingAnimator paging,
            ReaderContentController content,
            com.metahumanz.pacilread.reader.modern.tts.ReaderTtsController tts
    ) {
        this.chrome = chrome;
        this.paging = paging;
        this.content = content;
        this.tts = tts;
    }

    public void applyReaderSettings() {
        ReaderThemePalette palette = ReaderDisplayModeHelper.resolvePalette(activity, runtime.settingsStore);
        Typeface bodyTypeface = buildReaderTypeface(
                runtime.settingsStore.getReaderFontFamily(),
                runtime.settingsStore.getReaderFontWeight()
        );
        Typeface titleTypeface = buildReaderTypeface(
                runtime.settingsStore.getReaderFontFamily(),
                Math.max(600, Math.min(900, runtime.settingsStore.getReaderFontWeight() + 200))
        );
        applyDoublePageVisibility();
        int resolvedTextColor = resolveReaderTextColor(palette);
        state.currentReaderPageColor = palette.pageColor;
        state.currentReaderTextColor = resolvedTextColor;
        views.readerRoot.setBackgroundColor(palette.backgroundColor);
        chrome.applyReaderMenuPalette(palette, resolvedTextColor);
        views.pageCurrent.setBackgroundColor(android.graphics.Color.TRANSPARENT);
        views.pageIncoming.setBackgroundColor(android.graphics.Color.TRANSPARENT);

        int leftPadding = ui.dp(runtime.settingsStore.getLeftPaddingDp());
        int rightPadding = ui.dp(runtime.settingsStore.getRightPaddingDp());
        int topPadding = ui.dp(runtime.settingsStore.getTopPaddingDp())
                + state.readerContentInsetTop
                + computeHudReservedTopPx();
        int bottomPadding = ui.dp(runtime.settingsStore.getBottomPaddingDp())
                + state.readerContentInsetBottom
                + computeHudReservedBottomPx();
        ((ViewGroup) views.pageCurrent).setPadding(leftPadding, topPadding, rightPadding, bottomPadding);
        ((ViewGroup) views.pageIncoming).setPadding(leftPadding, topPadding, rightPadding, bottomPadding);

        stylePageTitleView(views.pageTitleCurrent, titleTypeface, resolvedTextColor);
        stylePageTitleView(views.pageTitleCurrentRight, titleTypeface, resolvedTextColor);
        stylePageTitleView(views.pageTitleIncoming, titleTypeface, resolvedTextColor);
        stylePageTitleView(views.pageTitleIncomingRight, titleTypeface, resolvedTextColor);
        stylePageBodyView(views.pageBodyCurrent, bodyTypeface, resolvedTextColor);
        stylePageBodyView(views.pageBodyCurrentRight, bodyTypeface, resolvedTextColor);
        stylePageBodyView(views.pageBodyIncoming, bodyTypeface, resolvedTextColor);
        stylePageBodyView(views.pageBodyIncomingRight, bodyTypeface, resolvedTextColor);
        styleHudTextView(views.hudTopLeft, bodyTypeface, resolvedTextColor);
        styleHudTextView(views.hudTopCenter, bodyTypeface, resolvedTextColor);
        styleHudTextView(views.hudTopRight, bodyTypeface, resolvedTextColor);
        styleHudTextView(views.hudBottomLeft, bodyTypeface, resolvedTextColor);
        styleHudTextView(views.hudBottomCenter, bodyTypeface, resolvedTextColor);
        styleHudTextView(views.hudBottomRight, bodyTypeface, resolvedTextColor);

        String alignment = runtime.settingsStore.getChapterTitleAlignment();
        int titleGravity = "center".equals(alignment) ? Gravity.CENTER : Gravity.LEFT;
        views.pageTitleCurrent.setGravity(titleGravity);
        views.pageTitleCurrentRight.setGravity(titleGravity);
        views.pageTitleIncoming.setGravity(titleGravity);
        views.pageTitleIncomingRight.setGravity(titleGravity);

        paging.invalidatePreparedPagingSnapshots();
        if (tts != null) {
            tts.updateTtsHighlight();
        }
        if (runtime.settingsStore.isKeepScreenOn()) {
            activity.getWindow().addFlags(WindowManager.LayoutParams.FLAG_KEEP_SCREEN_ON);
        } else {
            activity.getWindow().clearFlags(WindowManager.LayoutParams.FLAG_KEEP_SCREEN_ON);
        }
        applyBackgroundImage();
        chrome.updateReaderLayoutInsets();
        views.pageCurrent.requestLayout();
        views.pageIncoming.requestLayout();
        views.pageBodyCurrent.requestLayout();
        views.pageBodyCurrentRight.requestLayout();
        views.pageBodyIncoming.requestLayout();
        views.pageBodyIncomingRight.requestLayout();
        chrome.updateSystemBarsVisibility(state.controlsVisible);
        chrome.applyGlassOpacity();
        chrome.updateReaderHud();
        paging.invalidatePreparedPagingSnapshots();
    }

    public void attachBackground(Uri uri) {
        runtime.safeExecute(() -> {
            try {
                if (!activity.isReaderActive()) {
                    return;
                }
                String oldPath = runtime.settingsStore.getReaderBackgroundPath();
                File newFile = FileAssetHelper.copyUriToFolder(activity, uri, "backgrounds", "reader_bg");
                if (oldPath != null && !oldPath.isBlank()) {
                    FileAssetHelper.deleteIfExists(oldPath);
                }
                runtime.settingsStore.setReaderBackgroundPath(newFile.getAbsolutePath());
                activity.runOnReaderUiThread(this::applyReaderSettings);
            } catch (Exception error) {
                activity.runOnReaderUiThread(() -> ui.showToast("设置背景失败: " + error.getMessage()));
            }
        }, "attach reader background");
    }

    public void applyBackgroundImage() {
        String path = runtime.settingsStore.getReaderBackgroundPath();
        ReaderThemePalette palette = ReaderDisplayModeHelper.resolvePalette(activity, runtime.settingsStore);
        if (!shouldUseCustomBackground(path)) {
            int builtInRes = palette.backgroundDrawableRes;
            if (builtInRes != 0) {
                views.readerBackgroundImage.setImageResource(builtInRes);
                views.readerBackgroundImage.setVisibility(View.VISIBLE);
                applyBackgroundBlur();
            } else {
                views.readerBackgroundImage.setImageDrawable(null);
                views.readerBackgroundImage.setVisibility(View.GONE);
            }
            return;
        }
        Bitmap bitmap = BitmapFactory.decodeFile(path);
        if (bitmap == null) {
            int builtInRes = palette.backgroundDrawableRes;
            if (builtInRes != 0) {
                views.readerBackgroundImage.setImageResource(builtInRes);
                views.readerBackgroundImage.setVisibility(View.VISIBLE);
                applyBackgroundBlur();
            } else {
                views.readerBackgroundImage.setImageDrawable(null);
                views.readerBackgroundImage.setVisibility(View.GONE);
            }
            return;
        }
        views.readerBackgroundImage.setImageBitmap(bitmap);
        views.readerBackgroundImage.setVisibility(View.VISIBLE);
        applyBackgroundBlur();
    }

    public void applyBackgroundBlur() {
        int blurPercent = runtime.settingsStore.getBackgroundBlurPercent();
        if (blurPercent <= 0) {
            views.readerBackgroundImage.setAlpha(1.0f);
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.S) {
                views.readerBackgroundImage.setRenderEffect(null);
            }
            return;
        }
        float alpha = 1.0f - (blurPercent / 100f * 0.5f);
        views.readerBackgroundImage.setAlpha(alpha);
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.S) {
            float radius = blurPercent / 100f * 25f;
            android.graphics.RenderEffect blurEffect = android.graphics.RenderEffect.createBlurEffect(
                    radius,
                    radius,
                    android.graphics.Shader.TileMode.CLAMP
            );
            views.readerBackgroundImage.setRenderEffect(blurEffect);
        }
    }

    public void openBackgroundPicker(int requestCode) {
        Intent intent = new Intent(Intent.ACTION_OPEN_DOCUMENT);
        intent.addCategory(Intent.CATEGORY_OPENABLE);
        intent.setType("image/*");
        activity.startActivityForResult(intent, requestCode);
    }

    public String currentBackgroundLabel() {
        String path = runtime.settingsStore.getReaderBackgroundPath();
        ReaderThemePalette palette = ReaderDisplayModeHelper.resolvePalette(activity, runtime.settingsStore);
        if (!shouldUseCustomBackground(path)) {
            String suffix = ReaderDisplayModeHelper.isAutoNightActive(activity, runtime.settingsStore)
                    ? "（自动夜航）"
                    : "";
            return "当前背景：使用" + palette.displayName + "内置壁纸" + suffix;
        }
        return "当前背景：" + new File(path).getName();
    }

    public int resolveReaderTextColor(ReaderThemePalette palette) {
        String colorKey = runtime.settingsStore.getReaderTextColor();
        if (ReaderDisplayModeHelper.shouldOverrideCustomVisuals(activity, runtime.settingsStore, null)) {
            colorKey = "theme_default";
        }
        return resolveReaderTextColorValue(colorKey, palette);
    }

    public int resolveReaderTextColorValue(String colorKey, ReaderThemePalette palette) {
        if ("custom".equals(colorKey)) {
            String customColor = runtime.settingsStore.getCustomTextColor();
            if (customColor != null && !customColor.isEmpty()) {
                try {
                    return android.graphics.Color.parseColor(customColor);
                } catch (Exception ignore) {
                }
            }
            return palette.textColor;
        }
        if ("ink_brown".equals(colorKey)) {
            return 0xFF5A4330;
        }
        if ("graphite".equals(colorKey)) {
            return 0xFF374151;
        }
        if ("warm_gray".equals(colorKey)) {
            return 0xFF635B52;
        }
        if ("jade_ink".equals(colorKey)) {
            return 0xFF255B57;
        }
        if ("forest_ink".equals(colorKey)) {
            return 0xFF274235;
        }
        if ("moon_white".equals(colorKey)) {
            return 0xFFF5F7FA;
        }
        return palette.textColor;
    }

    public void updateTextColorPreview(TextView preview, String colorKey, ReaderThemePalette palette) {
        if (preview == null) {
            return;
        }
        int index = ReaderOptionCatalog.indexOf(ReaderOptionCatalog.READER_TEXT_COLOR_KEYS, colorKey, 0);
        preview.setText("字色预览：" + ReaderOptionCatalog.READER_TEXT_COLOR_LABELS[index]);
        preview.setTextColor(resolveReaderTextColorValue(colorKey, palette));
        preview.setBackgroundColor(palette.pageColor);
    }

    private int computeHudReservedTopPx() {
        boolean showCenterSlots = isLandscapeHudMode();
        if (!hasVisibleHudSlot(
                runtime.settingsStore.getHudTopLeft(),
                showCenterSlots ? runtime.settingsStore.getHudTopCenter() : "none",
                runtime.settingsStore.getHudTopRight()
        )) {
            return 0;
        }
        return ui.dp(runtime.settingsStore.getHudTopMarginDp() + HUD_BAR_HEIGHT_DP);
    }

    private int computeHudReservedBottomPx() {
        boolean showCenterSlots = isLandscapeHudMode();
        if (!hasVisibleHudSlot(
                runtime.settingsStore.getHudBottomLeft(),
                showCenterSlots ? runtime.settingsStore.getHudBottomCenter() : "none",
                runtime.settingsStore.getHudBottomRight()
        )) {
            return 0;
        }
        return ui.dp(runtime.settingsStore.getHudBottomMarginDp() + HUD_BAR_HEIGHT_DP);
    }

    private boolean isLandscapeHudMode() {
        return views.pageStage != null
                && views.pageStage.getWidth() > 0
                && views.pageStage.getHeight() > 0
                && views.pageStage.getWidth() > views.pageStage.getHeight();
    }

    private boolean hasVisibleHudSlot(String... slots) {
        if (slots == null) {
            return false;
        }
        for (String slot : slots) {
            if (slot != null && !"none".equals(slot)) {
                return true;
            }
        }
        return false;
    }

    private void applyDoublePageVisibility() {
        boolean active = ReaderDisplayModeHelper.isDoublePageActive(
                activity,
                runtime.settingsStore,
                views.pageStage == null ? 0 : views.pageStage.getWidth(),
                views.pageStage == null ? 0 : views.pageStage.getHeight()
        );
        int visibility = active ? View.VISIBLE : View.GONE;
        views.pageCurrentRightPane.setVisibility(visibility);
        views.pageIncomingRightPane.setVisibility(visibility);
        views.pageCurrentGutter.setVisibility(visibility);
        views.pageIncomingGutter.setVisibility(visibility);
    }

    private boolean shouldUseCustomBackground(String path) {
        if (path == null || path.isBlank()) {
            return false;
        }
        return !ReaderDisplayModeHelper.shouldOverrideCustomVisuals(activity, runtime.settingsStore, null);
    }

    private void stylePageTitleView(TextView textView, Typeface typeface, int textColor) {
        if (textView == null) {
            return;
        }
        textView.setTextColor(textColor);
        textView.setTypeface(typeface);
        textView.setIncludeFontPadding(false);
        textView.setTextSize(runtime.settingsStore.getFontSizeSp() * CHAPTER_TITLE_SCALE);
    }

    private void stylePageBodyView(com.metahumanz.pacilread.reader.JustifiedPageTextView textView, Typeface typeface, int textColor) {
        if (textView == null) {
            return;
        }
        textView.setTextColor(textColor);
        textView.setTypeface(typeface);
        textView.setTextSize(runtime.settingsStore.getFontSizeSp());
        textView.setLineSpacing(runtime.settingsStore.getLineSpacingExtraSp(), 1f);
        textView.setLetterSpacing(runtime.settingsStore.getLetterSpacing());
        textView.setFullJustifyEnabled(runtime.settingsStore.isBodyTextJustified());
        textView.setGravity(Gravity.START | Gravity.TOP);
        textView.setPadding(0, 0, 0, 0);
    }

    private void styleHudTextView(TextView textView, Typeface typeface, int textColor) {
        if (textView == null) {
            return;
        }
        textView.setTypeface(typeface);
        textView.setIncludeFontPadding(false);
        textView.setTextSize(HUD_TEXT_SIZE_SP);
        textView.setTextColor(applyAlpha(textColor, HUD_TEXT_ALPHA));
    }

    private int applyAlpha(int color, float alpha) {
        return Color.argb(
                Math.round(Color.alpha(color) * alpha),
                Color.red(color),
                Color.green(color),
                Color.blue(color)
        );
    }

    public void updateLetterSpacingLabel(TextView label, android.widget.SeekBar seekBar) {
        float spacing = seekBar.getProgress() / 20f;
        label.setText(String.format(Locale.SIMPLIFIED_CHINESE, "%.2f", spacing));
    }

    public void updateFirstLineIndentLabel(TextView label, android.widget.SeekBar seekBar) {
        label.setText(seekBar.getProgress() + " 字符");
    }

    public void updateParagraphSpacingLabel(TextView label, android.widget.SeekBar seekBar) {
        label.setText(seekBar.getProgress() + " dp");
    }

    public void updateBackgroundBlurLabel(TextView label, android.widget.SeekBar seekBar) {
        label.setText(seekBar.getProgress() + "%");
    }

    public Typeface buildReaderTypeface(String familyKey, int weight) {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.P) {
            return Typeface.create(familyKey, weight);
        }
        return Typeface.create(familyKey, weight >= 600 ? Typeface.BOLD : Typeface.NORMAL);
    }
}

File: app/src/main/kotlin/com/metahumanz/pacilread/reader/modern/ui/ReaderStyleController.kt
package com.metahumanz.pacilread.reader.modern.ui

import android.content.Intent
import android.graphics.BitmapFactory
import android.graphics.Color
import android.graphics.Typeface
import android.graphics.drawable.Drawable
import android.graphics.drawable.GradientDrawable
import android.net.Uri
import android.os.Build
import android.view.Gravity
import android.view.View
import android.view.ViewGroup
import android.view.WindowManager
import android.widget.SeekBar
import android.widget.TextView
import com.metahumanz.pacilread.reader.JustifiedPageTextView
import com.metahumanz.pacilread.reader.modern.ModernReaderActivity
import com.metahumanz.pacilread.reader.modern.ReaderRuntime
import com.metahumanz.pacilread.reader.modern.ReaderSessionState
import com.metahumanz.pacilread.reader.modern.ReaderUiUtils
import com.metahumanz.pacilread.reader.modern.ReaderViewRefs
import com.metahumanz.pacilread.reader.modern.config.ReaderOptionCatalog
import com.metahumanz.pacilread.reader.modern.content.ReaderContentController
import com.metahumanz.pacilread.reader.modern.paging.ReaderPagingAnimator
import com.metahumanz.pacilread.reader.modern.theme.ReaderDisplayModeHelper
import com.metahumanz.pacilread.reader.modern.theme.ReaderThemePalette
import com.metahumanz.pacilread.reader.modern.tts.ReaderTtsController
import com.metahumanz.pacilread.util.FileAssetHelper
import java.io.File
import java.util.Locale
import kotlin.math.max
import kotlin.math.min
import kotlin.math.pow
import kotlin.math.roundToInt

class ReaderStyleController(
    private val activity: ModernReaderActivity,
    private val runtime: ReaderRuntime,
    private val views: ReaderViewRefs,
    private val state: ReaderSessionState,
    private val ui: ReaderUiUtils,
) {
    private var chrome: ReaderChromeController? = null
    private var paging: ReaderPagingAnimator? = null
    @Suppress("unused")
    private var content: ReaderContentController? = null
    private var tts: ReaderTtsController? = null

    fun attachControllers(
        chrome: ReaderChromeController?,
        paging: ReaderPagingAnimator?,
        content: ReaderContentController?,
        tts: ReaderTtsController?,
    ) {
        this.chrome = chrome
        this.paging = paging
        this.content = content
        this.tts = tts
    }

    fun applyReaderSettings() {
        val attachedChrome = chrome!!
        val attachedPaging = paging!!
        val palette = ReaderDisplayModeHelper.resolvePalette(activity, runtime.settingsStore)
        val bodyTypeface = buildReaderTypeface(
            runtime.settingsStore.readerFontFamily,
            runtime.settingsStore.readerFontWeight,
        )
        val titleTypeface = buildReaderTypeface(
            runtime.settingsStore.readerFontFamily,
            max(600, min(900, runtime.settingsStore.readerFontWeight + 200)),
        )
        val resolvedTextColor = resolveReaderTextColor(palette)
        state.currentReaderPageColor = palette.pageColor
        state.currentReaderTextColor = resolvedTextColor
        applyDoublePageVisibility()
        views.readerRoot.setBackgroundColor(palette.backgroundColor)
        attachedChrome.applyReaderMenuPalette(palette, resolvedTextColor)
        views.pageCurrent.setBackgroundColor(Color.TRANSPARENT)
        views.pageIncoming.setBackgroundColor(Color.TRANSPARENT)

        val leftPadding = ui.dp(runtime.settingsStore.leftPaddingDp)
        val rightPadding = ui.dp(runtime.settingsStore.rightPaddingDp)
        val topPadding = ui.dp(runtime.settingsStore.topPaddingDp) +
            state.readerContentInsetTop + computeHudReservedTopPx()
        val bottomPadding = ui.dp(runtime.settingsStore.bottomPaddingDp) +
            state.readerContentInsetBottom + computeHudReservedBottomPx()
        (views.pageCurrent as ViewGroup).setPadding(leftPadding, topPadding, rightPadding, bottomPadding)
        (views.pageIncoming as ViewGroup).setPadding(leftPadding, topPadding, rightPadding, bottomPadding)

        stylePageTitleView(views.pageTitleCurrent, titleTypeface, resolvedTextColor)
        stylePageTitleView(views.pageTitleCurrentRight, titleTypeface, resolvedTextColor)
        stylePageTitleView(views.pageTitleIncoming, titleTypeface, resolvedTextColor)
        stylePageTitleView(views.pageTitleIncomingRight, titleTypeface, resolvedTextColor)
        stylePageBodyView(views.pageBodyCurrent, bodyTypeface, resolvedTextColor)
        stylePageBodyView(views.pageBodyCurrentRight, bodyTypeface, resolvedTextColor)
        stylePageBodyView(views.pageBodyIncoming, bodyTypeface, resolvedTextColor)
        stylePageBodyView(views.pageBodyIncomingRight, bodyTypeface, resolvedTextColor)
        styleHudTextView(views.hudTopLeft, bodyTypeface, resolvedTextColor)
        styleHudTextView(views.hudTopCenter, bodyTypeface, resolvedTextColor)
        styleHudTextView(views.hudTopRight, bodyTypeface, resolvedTextColor)
        styleHudTextView(views.hudBottomLeft, bodyTypeface, resolvedTextColor)
        styleHudTextView(views.hudBottomCenter, bodyTypeface, resolvedTextColor)
        styleHudTextView(views.hudBottomRight, bodyTypeface, resolvedTextColor)

        val titleGravity = if (runtime.settingsStore.chapterTitleAlignment == "center") Gravity.CENTER else Gravity.LEFT
        views.pageTitleCurrent.gravity = titleGravity
        views.pageTitleCurrentRight.gravity = titleGravity
        views.pageTitleIncoming.gravity = titleGravity
        views.pageTitleIncomingRight.gravity = titleGravity

        attachedPaging.invalidatePreparedPagingSnapshots()
        tts?.updateTtsHighlight()
        if (runtime.settingsStore.isKeepScreenOn) {
            activity.window.addFlags(WindowManager.LayoutParams.FLAG_KEEP_SCREEN_ON)
        } else {
            activity.window.clearFlags(WindowManager.LayoutParams.FLAG_KEEP_SCREEN_ON)
        }
        applyBackgroundImage()
        attachedChrome.updateReaderLayoutInsets()
        views.pageCurrent.requestLayout()
        views.pageIncoming.requestLayout()
        views.pageBodyCurrent.requestLayout()
        views.pageBodyCurrentRight.requestLayout()
        views.pageBodyIncoming.requestLayout()
        views.pageBodyIncomingRight.requestLayout()
        attachedChrome.updateSystemBarsVisibility(state.controlsVisible)
        attachedChrome.applyGlassOpacity()
        attachedChrome.updateReaderHud()
        attachedPaging.invalidatePreparedPagingSnapshots()
    }

    fun attachBackground(uri: Uri?) {
        runtime.safeExecute(Runnable {
            try {
                if (!activity.isReaderActive) return@Runnable
                val oldPath = runtime.settingsStore.readerBackgroundPath
                val newFile = FileAssetHelper.copyUriToFolder(activity, uri!!, "backgrounds", "reader_bg")
                if (!oldPath.isNullOrBlank()) FileAssetHelper.deleteIfExists(oldPath)
                runtime.settingsStore.readerBackgroundPath = newFile.absolutePath
                activity.runOnReaderUiThread(::applyReaderSettings)
            } catch (error: Exception) {
                activity.runOnReaderUiThread { ui.showToast("设置背景失败: ${error.message}") }
            }
        }, "attach reader background")
    }

    fun applyBackgroundImage() {
        val path = runtime.settingsStore.readerBackgroundPath
        val palette = ReaderDisplayModeHelper.resolvePalette(activity, runtime.settingsStore)
        if (!shouldUseCustomBackground(path)) {
            applyBuiltInBackground(palette)
            return
        }
        val bitmap = BitmapFactory.decodeFile(path)
        if (bitmap == null) {
            applyBuiltInBackground(palette)
            return
        }
        views.readerBackgroundImage.setImageBitmap(bitmap)
        views.readerBackgroundImage.visibility = View.VISIBLE
        applyBackgroundBlur()
    }

    private fun applyBuiltInBackground(palette: ReaderThemePalette) {
        if (palette.backgroundDrawableRes != 0) {
            views.readerBackgroundImage.setImageResource(palette.backgroundDrawableRes)
            views.readerBackgroundImage.visibility = View.VISIBLE
            applyBackgroundBlur()
        } else {
            views.readerBackgroundImage.setImageDrawable(null)
            views.readerBackgroundImage.visibility = View.GONE
        }
    }

    fun applyBackgroundBlur() {
        val blurPercent = runtime.settingsStore.backgroundBlurPercent
        if (blurPercent <= 0) {
            views.readerBackgroundImage.alpha = 1f
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.S) views.readerBackgroundImage.setRenderEffect(null)
            return
        }
        views.readerBackgroundImage.alpha = 1f - blurPercent / 100f * 0.5f
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.S) {
            val radius = blurPercent / 100f * 25f
            views.readerBackgroundImage.setRenderEffect(
                android.graphics.RenderEffect.createBlurEffect(
                    radius,
                    radius,
                    android.graphics.Shader.TileMode.CLAMP,
                ),
            )
        }
    }

    @Suppress("DEPRECATION")
    fun openBackgroundPicker(requestCode: Int) {
        val intent = Intent(Intent.ACTION_OPEN_DOCUMENT).apply {
            addCategory(Intent.CATEGORY_OPENABLE)
            type = "image/*"
        }
        activity.startActivityForResult(intent, requestCode)
    }

    fun currentBackgroundLabel(): String {
        val path = runtime.settingsStore.readerBackgroundPath
        val palette = ReaderDisplayModeHelper.resolvePalette(activity, runtime.settingsStore)
        if (!shouldUseCustomBackground(path)) {
            val suffix = if (ReaderDisplayModeHelper.isAutoNightActive(activity, runtime.settingsStore)) "（自动夜航）" else ""
            return "当前背景：使用${palette.displayName}内置壁纸$suffix"
        }
        return "当前背景：${File(path!!).name}"
    }

    fun resolveReaderTextColor(palette: ReaderThemePalette): Int {
        var colorKey = runtime.settingsStore.readerTextColor
        if (ReaderDisplayModeHelper.shouldOverrideCustomVisuals(activity, runtime.settingsStore, null)) {
            colorKey = "theme_default"
        }
        return resolveReaderTextColorValue(colorKey, palette)
    }

    fun resolveReaderTextColorValue(colorKey: String?, palette: ReaderThemePalette): Int {
        if (colorKey == "custom") {
            val customColor = runtime.settingsStore.customTextColor
            if (!customColor.isNullOrEmpty()) {
                try {
                    return Color.parseColor(customColor)
                } catch (_: Exception) {
                }
            }
            return palette.textColor
        }
        return when (colorKey) {
            "ink_brown" -> 0xFF5A4330.toInt()
            "graphite" -> 0xFF374151.toInt()
            "warm_gray" -> 0xFF635B52.toInt()
            "jade_ink" -> 0xFF255B57.toInt()
            "forest_ink" -> 0xFF274235.toInt()
            "moon_white" -> 0xFFF5F7FA.toInt()
            else -> palette.textColor
        }
    }

    fun updateTextColorPreview(preview: TextView?, colorKey: String?, palette: ReaderThemePalette) {
        preview ?: return
        val index = ReaderOptionCatalog.indexOf(ReaderOptionCatalog.READER_TEXT_COLOR_KEYS, colorKey, 0)
        preview.text = "字色预览：${ReaderOptionCatalog.READER_TEXT_COLOR_LABELS[index]}"
        preview.setTextColor(resolveReaderTextColorValue(colorKey, palette))
        preview.setBackgroundColor(palette.pageColor)
    }

    private fun computeHudReservedTopPx(): Int {
        val showCenterSlots = isLandscapeHudMode()
        if (!hasVisibleHudSlot(
                runtime.settingsStore.hudTopLeft,
                if (showCenterSlots) runtime.settingsStore.hudTopCenter else "none",
                runtime.settingsStore.hudTopRight,
            )) return 0
        return ui.dp(runtime.settingsStore.hudTopMarginDp + HUD_BAR_HEIGHT_DP)
    }

    private fun computeHudReservedBottomPx(): Int {
        val showCenterSlots = isLandscapeHudMode()
        if (!hasVisibleHudSlot(
                runtime.settingsStore.hudBottomLeft,
                if (showCenterSlots) runtime.settingsStore.hudBottomCenter else "none",
                runtime.settingsStore.hudBottomRight,
            )) return 0
        return ui.dp(runtime.settingsStore.hudBottomMarginDp + HUD_BAR_HEIGHT_DP)
    }

    private fun isLandscapeHudMode(): Boolean =
        views.pageStage.width > 0 && views.pageStage.height > 0 && views.pageStage.width > views.pageStage.height

    private fun hasVisibleHudSlot(vararg slots: String?): Boolean = slots.any { it != null && it != "none" }

    private fun applyDoublePageVisibility() {
        val active = ReaderDisplayModeHelper.isDoublePageActive(
            activity,
            runtime.settingsStore,
            views.pageStage.width,
            views.pageStage.height,
        )
        val visibility = if (active) View.VISIBLE else View.GONE
        views.pageCurrentRightPane.visibility = visibility
        views.pageIncomingRightPane.visibility = visibility
        val gutterVisibility = if (active && shouldShowDoublePageGutter()) View.VISIBLE else View.GONE
        views.pageCurrentGutter.visibility = gutterVisibility
        views.pageIncomingGutter.visibility = gutterVisibility
        val showOuterPageSimulationSpine = active &&
            runtime.settingsStore.flipMode == "simulation" &&
            runtime.settingsStore.simulationDoublePageTurnMode == "outerPage"
        if (showOuterPageSimulationSpine) {
            views.pageBookSpineOverlay.background = buildAdaptiveBookSpineDrawable(state.currentReaderPageColor)
        }
        views.pageBookSpineOverlay.visibility = if (showOuterPageSimulationSpine) View.VISIBLE else View.GONE
    }

    private fun shouldShowDoublePageGutter(): Boolean = true

    private fun buildAdaptiveBookSpineDrawable(pageColor: Int): Drawable {
        val lightInk = shouldUseLightSpineInk(pageColor)
        return GradientDrawable().apply { setColor(spineInkColor(lightInk, if (lightInk) 184 else 132)) }
    }

    private fun shouldUseLightSpineInk(pageColor: Int): Boolean = relativeLuminance(pageColor) < 0.45f

    private fun spineInkColor(lightInk: Boolean, alpha: Int): Int =
        if (lightInk) Color.argb(alpha, 255, 255, 255) else Color.argb(alpha, 0, 0, 0)

    private fun relativeLuminance(color: Int): Float = (
        0.2126 * linearizedChannel(Color.red(color)) +
            0.7152 * linearizedChannel(Color.green(color)) +
            0.0722 * linearizedChannel(Color.blue(color))
        ).toFloat()

    private fun linearizedChannel(value: Int): Double {
        val channel = value / 255.0
        return if (channel <= 0.03928) channel / 12.92 else ((channel + 0.055) / 1.055).pow(2.4)
    }

    private fun shouldUseCustomBackground(path: String?): Boolean =
        !path.isNullOrBlank() && !ReaderDisplayModeHelper.shouldOverrideCustomVisuals(activity, runtime.settingsStore, null)

    private fun stylePageTitleView(textView: TextView?, typeface: Typeface, textColor: Int) {
        textView ?: return
        textView.setTextColor(textColor)
        textView.typeface = typeface
        textView.includeFontPadding = false
        textView.textSize = runtime.settingsStore.fontSizeSp * CHAPTER_TITLE_SCALE
    }

    private fun stylePageBodyView(textView: JustifiedPageTextView?, typeface: Typeface, textColor: Int) {
        textView ?: return
        textView.setTextColor(textColor)
        textView.typeface = typeface
        textView.textSize = runtime.settingsStore.fontSizeSp
        textView.setLineSpacing(runtime.settingsStore.lineSpacingExtraSp, 1f)
        textView.letterSpacing = runtime.settingsStore.letterSpacing
        textView.setFullJustifyEnabled(runtime.settingsStore.isBodyTextJustified)
        textView.gravity = Gravity.START or Gravity.TOP
        textView.setPadding(0, 0, 0, 0)
    }

    private fun styleHudTextView(textView: TextView?, typeface: Typeface, textColor: Int) {
        textView ?: return
        textView.typeface = typeface
        textView.includeFontPadding = false
        textView.textSize = HUD_TEXT_SIZE_SP
        textView.setTextColor(applyAlpha(textColor, HUD_TEXT_ALPHA))
    }

    private fun applyAlpha(color: Int, alpha: Float): Int = Color.argb(
        (Color.alpha(color) * alpha).roundToInt(),
        Color.red(color),
        Color.green(color),
        Color.blue(color),
    )

    fun updateLetterSpacingLabel(label: TextView, seekBar: SeekBar) {
        label.text = String.format(Locale.SIMPLIFIED_CHINESE, "%.2f", seekBar.progress / 20f)
    }

    fun updateFirstLineIndentLabel(label: TextView, seekBar: SeekBar) {
        label.text = "${seekBar.progress} 字符"
    }

    fun updateParagraphSpacingLabel(label: TextView, seekBar: SeekBar) {
        label.text = "${seekBar.progress} dp"
    }

    fun updateBackgroundBlurLabel(label: TextView, seekBar: SeekBar) {
        label.text = "${seekBar.progress}%"
    }

    fun buildReaderTypeface(familyKey: String?, weight: Int): Typeface =
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.P) {
            Typeface.create(familyKey, weight)
        } else {
            Typeface.create(familyKey, if (weight >= 600) Typeface.BOLD else Typeface.NORMAL)
        }

    private companion object {
        const val CHAPTER_TITLE_SCALE = 1.4f
        const val HUD_BAR_HEIGHT_DP = 22
        const val HUD_TEXT_SIZE_SP = 12f
        const val HUD_TEXT_ALPHA = 0.78f
    }
}
